airbyte-cdk/python/airbyte_cdk/sources/declarative/requesters/retriers/default_retrier.py

```diff
@@ -2,27 +2,105 @@
+from airbyte_cdk.sources.declarative.requesters.retriers.backoff_strategies.exponential_backoff_strategy import ExponentialBackoffStrategy
+from airbyte_cdk.sources.declarative.requesters.retriers.backoff_strategy import BackoffStrategy
+from airbyte_cdk.sources.declarative.requesters.retriers.http_response_filter import HttpResponseFilter
+from airbyte_cdk.sources.declarative.requesters.retriers.retrier import (
+    NonRetriableResponseStatus,
+    ResponseStatus,
+    Retrier,
+    RetryResponseStatus,
+)
 
 
 class DefaultRetrier(Retrier):
-    def __init__(self, max_retries: Optional[int] = 5, retry_factor: float = 5):
+    """
+    Sample configs:
+
+    1. retry 10 times
+    `
+        retrier:
+          max_retries: 10
+    `
+    2. backoff for 5 seconds
+    `
+        retrier:
+          backoff_strategy:
+            - type: "ConstantBackoffStrategy"
+              backoff_time_in_seconds: 5
+    `
+    3. retry on HTTP 404
+    `
+        retrier:
+          retry_response_filter:
+            http_codes: [ 404 ]
+    `
+    4. ignore HTTP 404
+    `
+        retrier:
+          ignore_response_filter:
+            http_codes: [ 404 ]
+    `
+    5. retry if error message contains `retrythisrequest!` substring
+    `
+        retrier:
+          retry_response_filter:
+            error_message_contain: "retrythisrequest!"
+    `
+    6. retry if 'code' is a field present in the response body
+    `
+        retrier:
+          retry_response_filter:
+            predicate: "{{ 'code' in decoded_response }}"
+    `
+    """
+
+    DEFAULT_BACKOFF_STRATEGY = ExponentialBackoffStrategy
+
+    def __init__(
+        self,
+        retry_response_filter: HttpResponseFilter = None,
+        ignore_response_filter: HttpResponseFilter = None,
+        max_retries: Optional[int] = 5,
+        backoff_strategy: Optional[List[BackoffStrategy]] = None,
+    ):
         self._max_retries = max_retries
-        self._retry_factor = retry_factor
+        self._retry_response_filter = retry_response_filter or HttpResponseFilter(HttpResponseFilter.DEFAULT_RETRIABLE_ERRORS)
+        self._ignore_response_filter = ignore_response_filter or HttpResponseFilter(set())
+
+        if backoff_strategy:
+            self._backoff_strategy = backoff_strategy
+        else:
+            self._backoff_strategy = [DefaultRetrier.DEFAULT_BACKOFF_STRATEGY()]
+
+        self._last_request_to_attempt_count = {}
 
     @property
     def max_retries(self) -> Union[int, None]:
         return self._max_retries
 
-    @property
-    def retry_factor(self) -> float:
-        return self._retry_factor
-
-    def should_retry(self, response: requests.Response) -> bool:
-        return response.status_code == 429 or 500 <= response.status_code < 600
+    def should_retry(self, response: requests.Response) -> ResponseStatus:
+        url = response.request.url
+        if url not in self._last_request_to_attempt_count:
```

Comment:
this seems like it wouldn't work very well with POST requests right? since their distinguishing characteristic wouldn't be the URL address. And maybe an obvious question but this takes into account query params right?

Reply:
seems this is because backoff strategies are not stateful. Is there an issue with just making them stateful? May not be needed if we can figure something out for POSTs but apart from a hash of the request body/request in general, i"m not coming up with many ideas

Reply:
@sherifnada we can make them stateful, but won't that just move the check to the strategy instead of the retrier?

either way, good call on hashing the request. I'll run some tests to figure out exactly what can/should be used for comparison.

Reply:
@sherifnada there are at least 2 ways we could handle this
1. compare the requests content:
```
>>> hash(req2.method) == hash(req.method)
True
>>> hash(req2.url) == hash(req.url)
True
>>> req2.headers == req.headers
True
>>> hash(req2.body) == hash(req.body)
True
>>> hash(req2._body_position) == hash(req._body_position)
True
>>> req2._cookies == req._cookies
True
```
2. compare the requests for equality - [the retried prepared request is the same as the original one](https://github.com/airbytehq/airbyte/blob/master/airbyte-cdk/python/airbyte_cdk/sources/streams/http/http.py#L334)

I'm leaning towards option 2 because i think submitting a new request is semantically different from retrying one